src/workerd/io/actor-sqlite.c++

```diff
@@ -283,7 +406,28 @@ kj::Maybe<kj::Promise<void>> ActorSqlite::setAlarm(
     kj::Maybe<kj::Date> newAlarmTime, WriteOptions options) {
   requireNotBroken();
 
-  return hooks.setAlarm(newAlarmTime);
+  // TODO(soon): Need special logic to handle case where actor is using alarms without using
+  // other storage?
+
+  if (!haveDeferredDelete) {
+    // Skip setting alarm if alarm is already set to the given value.
+    //
+    // If we're in the alarm handler and haven't set the time yet, we can't perform this
+    // optimization as the current alarm time may be equal to the requested time but we indicate
+    // to the actor in getAlarm() that there is no alarm set, therefore we need to act like that
+    // in setAlarm().
+    //
+    // After the first write in the handler occurs, the logic here is correct again as the current
+    // alarm time would match what we are reporting to the user from getAlarm().
+    //
+    // TODO(now): optimization no longer useful in sqlite?
+    if (metadata.getAlarm() == newAlarmTime) {
+      return kj::none;
+    }
+  }
+  metadata.setAlarm(newAlarmTime);
+  haveDeferredDelete = false;
```

Comment:
Oh, good point... this does not work with transaction rollbacks.  I think to fix this we'd need more complex tracking of the deferred delete state -- e.g. only setting `haveDeferredDelete` to false when a transaction with an alarm change is committed.

It looks like `ActorCache` avoids this problem by explicitly tracking alarm changes in `Transaction` objects, and maybe the simplest solution might be to do something similar here.  It looks like the root `setAlarm()` method might have a similar problem though, or might interact weirdly with `txn.setAlarm()`?  I'd need to read more code to see if we do something to explicitly disallow use of the root `setAlarm()` method during transactions, or other unusual interactions between root and transaction-level methods.  As currently implemented in ActorSqlite, it looks like the transaction-level methods are all aliases for the root-level methods.

Reply:
> It looks like ActorCache avoids this problem by explicitly tracking alarm changes in Transaction objects, and maybe the simplest solution might be to do something similar here

Yeah, I think so.

> It looks like the root setAlarm() method might have a similar problem though

No, not as far as I can tell. It applies the change immediately to its in-memory state and immediately schedules a flush to storage along with an output gate, the same as any write in ActorCache.

> I'd need to read more code to see if we do something to explicitly disallow use of the root setAlarm() method during transactions, or other unusual interactions between root and transaction-level methods.

In the ActorCache model, there really weren't any interactions. Root methods apply directly and immediately even if there is an ongoing transaction, while changes in transaction get buffered up until commit time and only flushed then.

Reply:
> In the ActorCache model, there really weren't any interactions. Root methods apply directly and immediately even if there is an ongoing transaction, while changes in transaction get buffered up until commit time and only flushed then.

Ah, OK, then ActorCache doesn't have the problem I was thinking of.  I was confused by the fact ActorSqlite seems to have different semantics...  Its transaction methods are direct passthroughs to the root methods, such that any operation becomes part of the innermost nested open transaction.  So if someone did:
```
async alarm() {
  await this.storage.startTransaction((txn) => {
    this.storage.setAlarm(Date.now());
    txn.rollback();
  });
}
```
...for ActorSqlite, the `setAlarm()` would be part of the transaction, so the alarm state change would be rolled back, and the user might have some expectation that the alarm would be deleted at the end of the handler.  But for ActorCache, it sounds like the change would not be rolled back and the alarm would not be deleted at the end of the handler.

@kentonv I guess this difference in behavior might be intentional, to support other features, and would also happen for other methods like `this.storage.put()`?  If so, I assume it would be most useful for ActorSqlite to delete the alarm at the end of the handler in the above example, rather than maybe trying to match the ActorCache behavior.

Reply:
Yeah, under sqlite, when a transaction is open, there's no way for us to support doing queries that skip the transaction. And we also wanted transactions to include SQL queries with the new SQL interface. So we just said: ok, transactions now cover everything done within the transaction callback whether you issue the queries against the `txn` object or the underlying `storage` object. `txn` is essentially deprecated and people should stop declaring it as an argument.

Reply:
Added transaction alarm dirtiness tracking and test cases in 2d833464 .